From 7435bf7794ac7ef30f1d1e0612165fbebf0e8ff7 Mon Sep 17 00:00:00 2001
From: "lianghu.su" <lianghu.su@amlogic.com>
Date: Wed, 30 Jul 2014 14:49:57 +0800
Subject: [PATCH 4668/5965] PD# 88765:unplug hdmi_in cable, hdmi_in App quit
 need time about 10 seconds. by Wang Chengshun.

Change-Id: I089ba77fa06dc6ef3c457218d41f1c31fa956ad5

Conflicts:

	sound/soc/aml/m8/aml_i2s.c
---
 sound/soc/aml/m8/aml_i2s.c | 11 ++++++++++-
 sound/soc/aml/m8/aml_i2s.h |  1 +
 2 files changed, 11 insertions(+), 1 deletion(-)

diff --git a/sound/soc/aml/m8/aml_i2s.c b/sound/soc/aml/m8/aml_i2s.c
index 9b25ddab74f5..c32caa0d9df6 100755
--- a/sound/soc/aml/m8/aml_i2s.c
+++ b/sound/soc/aml/m8/aml_i2s.c
@@ -297,6 +297,7 @@ static int aml_i2s_trigger(struct snd_pcm_substream *substream,
 	  add_timer(&prtd->timer);
 #endif
 
+	  s->xrun_num = 0;
 	  s->active = 1;
 	  spin_unlock(&s->lock);
 	  break;		/* SNDRV_PCM_TRIGGER_START */
@@ -306,6 +307,7 @@ static int aml_i2s_trigger(struct snd_pcm_substream *substream,
 		// TODO
 	    spin_lock(&s->lock);
 	    s->active = 0;
+	    s->xrun_num = 0;
 	    spin_unlock(&s->lock);
 	    break;
 	default:
@@ -411,7 +413,7 @@ static void aml_i2s_timer_callback(unsigned long data)
     struct aml_runtime_data *prtd = runtime->private_data;
 		audio_stream_t *s = &prtd->s;
 
-    unsigned int last_ptr, size;
+    unsigned int last_ptr, size = 0;
 	if(substream->stream == SNDRV_PCM_STREAM_PLAYBACK)
 	{
 		if(s->active == 1)
@@ -451,6 +453,12 @@ static void aml_i2s_timer_callback(unsigned long data)
 				last_ptr = audio_in_spdif_wr_ptr();
 			if (last_ptr < s->last_ptr) {
 				size = runtime->dma_bytes + (last_ptr - (s->last_ptr))/2;
+			} else if (last_ptr == s->last_ptr) {
+				if (s->xrun_num++ > 100) {
+					printk(KERN_INFO "alsa capture long time no data, quit xrun!\n");
+					s->xrun_num = 0;
+					s->size = runtime->period_size;
+				}
 			} else {
 				size = (last_ptr - (s->last_ptr))/2;
 			}
@@ -541,6 +549,7 @@ static int aml_i2s_open(struct snd_pcm_substream *substream)
 #endif
 
 	spin_lock_init(&prtd->s.lock);
+	s->xrun_num = 0;
 	WRITE_MPEG_REG_BITS(MPLL_I2S_CNTL, 1,14, 1);
 	mutex_lock(&gate_mutex);
 	if(!num_clk_gate){
diff --git a/sound/soc/aml/m8/aml_i2s.h b/sound/soc/aml/m8/aml_i2s.h
index 564875b94bd1..919ba839ff76 100755
--- a/sound/soc/aml/m8/aml_i2s.h
+++ b/sound/soc/aml/m8/aml_i2s.h
@@ -19,6 +19,7 @@ typedef struct audio_stream {
     struct snd_pcm_substream *stream;
 	unsigned i2s_mode; //0:master, 1:slave,
     unsigned device_type;
+    unsigned int xrun_num;
 } audio_stream_t;
 
 typedef struct aml_audio {
-- 
2.19.0

